### ISSUE TITLE
inconsitency in oval declarations

### ISSUE THREAD

**Issue by passenger94** [Shoes3/shoes3#76]
manual says **oval(left, top, radius)** but its diameter not radius
while oval with style declaration is right
```ruby
Shoes.app {
  oval 0, 0, 50 # this is diameter not radius
  oval left: 50, top: 0, width: 50, height: 50
  oval left: 100, top: 0, radius: 50
  stroke red
  strokewidth 2
  line 150, 0, 150, 50
}
```

**Comment by passenger94** [comment]
in * shoes_canvas_oval*
https://github.com/passenger94/shoes3/blob/master/shoes/canvas.c#L514
> VALUE attr = shoes_shape_attr(argc, argv, 4, s_left, s_top, s_width, s_height);

doesn't this means that third argument to oval is in fact :width and not :radius
we should have **oval(left, top, width, height)** height being optional (tested)

**Comment by BackOrder** [comment]
Your assessment is correct. The parameters are processed correctly in **shoes_canvas_oval()** before passing on **shoes_add_shape()** in *canvas.c* — code below shows the values. It seems instead that the values are not correct before entering **shoes_shape_sketch()** in *ruby.c*.

SETUP macro might need investigation. https://github.com/Shoes3/shoes3/blob/master/shoes/ruby.c#L688

https://github.com/Shoes3/shoes3/blob/master/shoes/canvas.c#L517
```C
rb_warn("shoes_canvas_oval: %s\n", RSTRING_PTR(rb_inspect(attr)));
```
```
bug076.rb:2: warning: shoes_canvas_oval: { :left => 0, :top => 0, :width => 50 }

bug076.rb:3: warning: shoes_canvas_oval: { :left => 50, :top => 0, :width => 50, :height => 50 }

bug076.rb:4: warning: shoes_canvas_oval: { :left => 100, :top => 0, :radius => 50 }
```

```
shoes_shape_draw (self=41462712, c=41465760, actual=2) at shoes/ruby.c:1007
1007      self_t->place = place;
(gdb)
1008      return self;
(gdb) print place
$6 = {x = 0, y = 0, w = 50, h = 50, dx = 0, dy = 0, ix = 0, iy = 0, iw = 50,
  ih = 50, flags = 48 '0'}
```

**Print out on the faulty shape:**
```
shoes_shape_draw (self=41459808, c=41465760, actual=2) at shoes/ruby.c:1007
1007      self_t->place = place;
(gdb) print place
$15 = {x = 100, y = 0, w = 100, h = 100, dx = 0, dy = 0, ix = 100, iy = 0,
  iw = 100, ih = 100, flags = 48 '0'}
```


> doesn't this means that third argument to oval is in fact :width and not :radius
we should have oval(left, top, width, height) height being optional (tested)

Most likely radius would be converted to width and height. As you can see, internally, it is not setting **iw** and **ih** properly.

**Comment by BackOrder** [comment]
For posterity, a screenshot on Windows. @passenger94 don't forget that you can set labels about the platforms affected by this issue. You can set most of the labels if so you wish but **priorities** are usually set by @ccoupe since he is the official maintainer.

![image](https://cloud.githubusercontent.com/assets/8844578/6420957/b7d80a64-be98-11e4-8eee-651ed9ee923d.png)

**Comment by passenger94** [comment]
> bug076.rb:2: warning: shoes_canvas_oval: { :left => 0, :top => 0, :width => 50 }

so this means third argument is indeed :width not :radius !
the third shape is not faulty to me : with a radius of 50 you get a width of 100 (w = h = iw = ih = 100) 
unless i miss something ?

**Comment by BackOrder** [comment]
Yes, the missing something is that the named parameters are **styles** whereas `oval 0, 0, 50` are in-place method arguments.

**Comment by passenger94** [comment]
yes i do get this, but it seems, looking at other shape methods, that the named parameters are the styles choosen to appear as in-place arguments ! no ?

**Comment by BackOrder** [comment]
The manual makes it seems in order but it is for convenience. Named parameters is basically a hash and can be in any order.

**Comment by passenger94** [comment]
https://github.com/passenger94/shoes3/blob/master/shoes/ruby.c#L815
*shoes_shape_attr* takes as first argument the number of in-place arguments of **oval** method then takes the following symbols given to *shoes_shape_attr*  and add them, with corresponding values coming from in-place args of **oval** method, as key/value pair, to the hash eventually found as the last argument of **oval** method or newly created inside of *shoes_shape_attr*
if there is only a hash given to **oval** then this one is feeding *shoes_shape_attr*

this is how i understand *shoes_shape_attr*, wrong?

**Comment by BackOrder** [comment]
The problem is right here under our noses. The radius is multiplied by 2 in **ruby.c:shoes_place_exact()**. Removing the multiplication works with Oval shape but the function is used in few other occasions.

Would you mind testing the solution on oval, image and others shapes that may and may not be affected? The Shoes rendering should be the same on patched and unpatched versions except oval should now support radius properly.

https://github.com/Shoes3/shoes3/blob/master/shoes/ruby.c#L425

**SOLUTION**
```ruby
r = ATTR2(int, attr, radius, 0);
```

**Comment by passenger94** [comment]
no ! !!!

![ovals2](https://cloud.githubusercontent.com/assets/10730930/6425461/c9400c0c-bf29-11e4-9b12-c072eb12962f.jpg)

> bug076.rb:2: warning: shoes_canvas_oval: { :strokewidth => 2, :stroke => rgb(0, 0, 255), :fill => rgb(255, 255, 0), :left => 0, :top => 0, :width => 50, :height => 25 }

> bug076.rb:3: warning: shoes_canvas_oval: { :left => 50, :top => 0, :width => 50, :height => 50 }

> bug076.rb:4: warning: shoes_canvas_oval: { :left => 100, :top => 0, :radius => 50 }

third is of diameter 50 and radius 25 in reality ! because of the now missing multiplication by 2
the red line is 50px long ...

third parameter is **width** and fourth is **height** that's the *problem*
we only need to change the manual !

**Comment by BackOrder** [comment]
Seems like good work. Have you also tested when all those parameters are defined? For example, `oval 0, 0, 50, :width => 100`, `oval 0, 0, 50, :radius 10`, `oval :left => 0, :top => 0, :radius => 50`, `oval :left => 0, :top => 0, :radius => 50, :width => 100`, etc.

https://github.com/Shoes3/shoes3/blob/master/shoes/ruby.c#L425

My understanding is that width has higher priority than radius and height is replaced by width if it doesn't exist.

**Comment by passenger94** [comment]
yes tested with all kinds of combinations
in place arguments takes precedence over styles attributes because they are evaluated after the hash (hashes are ordered by order of entering the hash)
yes in case of styles attributes only, width takes precedence over radius and height is optional (takes the value of width) but if height only is used it is taken into account **with** radius drawing eventually an oval (not a circle)

one need to make a choice here : 
* manual is right and so correct c code in **shoes_canvas_oval** but this means old code breaks ie infamous *simple-sphere.rb*
* shoes_canvas_oval is right and so correct manual **oval(left, top, width, height)**
?

**Comment by BackOrder** [comment]
> manual is right and so correct c code in shoes_canvas_oval but this means old code breaks ie infamous simple-sphere.rb - doesn't crash shoes though -

Historically Curious release would take **x, y, w, h** for in-place parameters but will consider h = w when h not is provided and minimum of **:left, :top and :radius** for style parameters. The code we know today is based on Raisin release.

https://github.com/shoes/shoes/blob/1/Curious/shoes/canvas.c#L528 (Curious release)
https://github.com/shoes/shoes/blob/Raisins/shoes/ruby.c#L415 (Raisin release)

My understanding of the manual is that **left, top, radius** is correct but one could alternatively input **letf, top, width, height**. It is stated at the bottom of the entry in the manual. The number of parameters would tell the oval whether it is a radius or width/height. I'd vote for that.

**Comment by passenger94** [comment]
What is stated at the bottom of the oval entry is the real thing.
The best thing to do is probably changing radius for diameter in manual - which is what it is : diameter is equal to width (when height = width / 3 arguments) - and we're set ! (at the "cost" of having a sort of mismatch with :radius style ).

**Comment by BackOrder** [comment]
It shouldn't be a big deal to implement as the manual says. When you come to think about it, this is what the `r = ATTR2(int, attr, radius, 0) * 2;` is about. It multiplies by 2 to set width. There is not any real need to change to diameter as it can be fixed easily and comfort to the manual.

**Comment by passenger94** [comment]
**oval 0, 0, 50**
`r = ATTR2(int, attr, radius, 0) * 2; ` means give me radius if it exists, otherwise give me 0 and multiply by 2
so if third in-place arg is radius --> r = 100
`place->iw = place->w = ATTR2(int, attr, width, r);` so here w must be 100 ( no :width given)
examples proves we have 50 not 100 !

if third in-place arg is width --> r = 0
`place->iw = place->w = ATTR2(int, attr, width, r);` here w must be 50 (:width is 50)
as in examples.
also this is the place where :width takes precedence over :radius

`place->ih = place->h = ATTR2(int, attr, height, place->w);` if no height given, proceed with width

i can't see an easy way to accomodate manual here ?
simply getting rid of the multiplication ruins the radius style as we saw ...

**Comment by ccoupe** [comment]
Which one is consistent with the samples/good-clock.rb and good-follow.rb ? Those are more likely to be duplicated by users as they work, unlike the conflicting manual.

**Comment by BackOrder** [comment]
good-follow.rb uses radius and good-clock.rb uses width and height.

The code in Curious makes it clear that it had a fall through strategy to assign either one. This is absolutely possible in Federales and is in the mandate of maintenance. I don't have time to check on that over the weekend though.

**Comment by ccoupe** [comment]
Any progress on this issue? Solved? Will never be solved?

**Comment by passenger94** [comment]
one simple solution is to rewrite the oval entry in the manual, like :
>=== oval(left, top, width, height) » Shoes::Shape ===
>
>Draws a circular form of varied proportions at pixel coordinates (left, top) of width and height pixels, if >you omit the last argument you'll end up with a perfect circle of width pixels (think diameter).
>...

**Comment by BackOrder** [comment]
Guys, any progress on this one? We have a good idea of information in this thread but it suddenly stop at some point. Though it is obviously not a priority. Just catching up...
